File: python/13_Roman_to_Integer.py
class Solution:
    # @param {string} s
    # @return {integer}
    def romanToInt(self, s):
        if s is None or len(s)==0:
            return 0
        rtoi={
            'I':1,
            'V':5,
            'X':10,
            'L':50,
            'C':100,
            'D':500,
            'M':1000,
            }
        res = 0
        for i in range(len(s)):
            if i+1<len(s) and ((s[i]=='C' and s[i+1] in 'DM') or (s[i]=='X' and s[i+1] in 'LC') or (s[i]=='I' and s[i+1] in 'VX')):
                res -= rtoi[s[i]]
            else:
                res += rtoi[s[i]]
        return res

File: Java/13_Roman_to_Integer.java
public class Solution {
    public int romanToInt(String s) {
        if(s==null || s.length()==0){
            return 0;
        }
        char[] digits = s.toCharArray();
        int res=0, i=digits.length-1;
        char last='.';
        while(i>=0){
            switch(digits[i]){
                case 'I':
                    if(last!='.'&&(last=='V'||last=='X')){// there are two cases in total that we should substract the current digit.
                        res -= 1;
                    }else{
                        res += 1;
                    }
                    break;
                case 'V':
                    res += 5;
                    break;
                case 'X':
                    if(last!='.'&&last=='L'||last=='C'){// there are two cases in total that we should substract the current digit.
                        res -= 10;
                    }else{
                        res += 10;
                    }
                    break;
                case 'L':
                    res += 50;
                    break;
                case 'C':
                    if(last!='.'&&last=='D'||last=='M'){// there are two cases in total that we should substract the current digit.
                        res -= 100;
                    }else{
                        res += 100;
                    }
                    break;
                case 'D':
                    res += 500;
                    break;
                case 'M':
                    res += 1000;
                    break;
            }
            last = digits[i--];
        }
        return res;
    }
}